sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java

```diff
@@ -0,0 +1,681 @@
+
+        private List<HRegionLocation> getRegionLocations()
+                throws Exception {
+            final Scan scan = read.serializableScan.getScan();
+            byte[] startRow = scan.getStartRow();
+            byte[] stopRow = scan.getStopRow();
+
+            final List<HRegionLocation> regionLocations = new ArrayList<>();
+
+            final boolean scanWithNoLowerBound = startRow.length == 0;
+            final boolean scanWithNoUpperBound = stopRow.length == 0;
+            Configuration configuration = this.read.serializableConfiguration.getConfiguration();
+            try (Connection connection = ConnectionFactory.createConnection(configuration)) {
+                TableName tableName = TableName.valueOf(read.tableId);
+                RegionLocator regionLocator = connection.getRegionLocator(tableName);
+                List<HRegionLocation> tableRegionInfos = regionLocator.getAllRegionLocations();
+                for (HRegionLocation regionLocation : tableRegionInfos) {
+                    final byte[] startKey = regionLocation.getRegionInfo().getStartKey();
+                    final byte[] endKey = regionLocation.getRegionInfo().getEndKey();
+                    boolean isLastRegion = endKey.length == 0;
+                    // filters regions who are part of the scan
+                    if ((scanWithNoLowerBound
+                            || isLastRegion || Bytes.compareTo(startRow, endKey) < 0)
+                            && (scanWithNoUpperBound || Bytes.compareTo(stopRow, startKey) > 0)) {
+                        regionLocations.add(regionLocation);
+                    }
+                }
+            }
+            return regionLocations;
+        }
+
+        private List<HBaseSource>
+            splitBasedOnRegions(List<HRegionLocation> regionLocations, int numSplits)
+                throws Exception {
+            final Scan scan = read.serializableScan.getScan();
+            byte[] startRow = scan.getStartRow();
+            byte[] stopRow = scan.getStopRow();
+
+            final List<HBaseSource> sources = new ArrayList<>(numSplits);
+            final boolean scanWithNoLowerBound = startRow.length == 0;
+            final boolean scanWithNoUpperBound = stopRow.length == 0;
+
+            for (HRegionLocation regionLocation : regionLocations) {
+                final byte[] startKey = regionLocation.getRegionInfo().getStartKey();
+                final byte[] endKey = regionLocation.getRegionInfo().getEndKey();
+                boolean isLastRegion = endKey.length == 0;
+                String host = regionLocation.getHostnamePort();
+
+                final byte[] splitStart = (scanWithNoLowerBound
+                        || Bytes.compareTo(startKey, startRow) >= 0) ? startKey : startRow;
+                final byte[] splitStop =
+                        (scanWithNoUpperBound || Bytes.compareTo(endKey, stopRow) <= 0)
+                                && !isLastRegion ? endKey : stopRow;
```

Comment:
Is there a reason you didn't use the ByteKey and ByteKeyRange functions like overlap and contains to simplify this logic?